akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala

```diff
@@ -0,0 +1,865 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.javadsl
+
+import akka.event.LoggingAdapter
+import akka.japi.function
+import akka.stream._
+import akka.stream.impl.Stages.StageModule
+import akka.stream.impl.ConstantFun
+import akka.stream.stage.Stage
+import scala.collection.immutable
+import scala.collection.JavaConverters._
+import scala.annotation.unchecked.uncheckedVariance
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * A “stream of streams” sub-flow of data elements, e.g. produced by `groupBy`.
+ * SubFlows cannot contribute to the super-flow’s materialized value since they
+ * are materialized later, during the runtime of the flow graph processing.
+ */
+class SubFlow[-In, +Out, +Mat](delegate: scaladsl.SubFlow[Out, Mat, scaladsl.Flow[In, Out, Mat]#Repr, scaladsl.Sink[In, Mat]]) {
+
+  /** Converts this Flow to its Scala DSL counterpart */
+  def asScala: scaladsl.SubFlow[Out, Mat, scaladsl.Flow[In, Out, Mat]#Repr, scaladsl.Sink[In, Mat]] @uncheckedVariance = delegate
+
+  /**
+   * Flatten the sub-flows back into the super-flow, performing a merge of the
+   * given breadth. For concatenation of the flows use a breadth of 1 but beware
+   * that this will lead to deadlocks if the sub-flows need to make progress
+   * concurrently (as is generally the case when using `groupBy` on an unordered
+   * input stream).
+   *
+   * WARNING: currently only a breadth of 1 is actually implemented! Beware of deadlocks!
+   */
+  // FIXME remove doc warning once fixed
+  def mergeBack(breadth: Int): Flow[In, Out, Mat] =
```

Comment:
It's not `maximumSubstreams` though. It's `maxConcurrentSubstreams`

On Wed, Nov 25, 2015 at 12:24 PM, Roland Kuhn <notifications@github.com>
wrote:

> In akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala
> <https://github.com/akka/akka/pull/19006#discussion_r45855773>:
>
> > +class SubFlow[-In, +Out, +Mat](delegate: scaladsl.SubFlow[Out, Mat, scaladsl.Flow[In, Out, Mat]#Repr, scaladsl.Sink[In, Mat]]) {
> > +
> > +  /** Converts this Flow to its Scala DSL counterpart */
> > +  def asScala: scaladsl.SubFlow[Out, Mat, scaladsl.Flow[In, Out, Mat]#Repr, scaladsl.Sink[In, Mat]] @uncheckedVariance = delegate
> > +
> > +  /**
> > +   * Flatten the sub-flows back into the super-flow, performing a merge of the
> > +   * given breadth. For concatenation of the flows use a breadth of 1 but beware
> > +   * that this will lead to deadlocks if the sub-flows need to make progress
> > +   * concurrently (as is generally the case when using `groupBy` on an unordered
> > +   * input stream).
> > +   *
> > +   * WARNING: currently only a breadth of 1 is actually implemented! Beware of deadlocks!
> > +   */
> > +  // FIXME remove doc warning once fixed
> > +  def mergeBack(breadth: Int): Flow[In, Out, Mat] =
>
> Okay, so here’s what I’ll do:
>
>    - remove the parameter and rename to mergeSubstreams
>    - add concatSubstreams method
>    - add maximumSubstreams parameter to groupBy
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19006/files#r45855773>.
>



-- 
Cheers,
√